[c10d] allow nonblocking wrap of ncclCommInitRankConfig

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118256

resolve #117749 

Summary:
Updated the PR with the following intentions:

1. identify eagerMode init (as opposed to lazy init), in which case we will create NCCL comms without guarantees that they are fully initialized if NONBLOCKING mode is also enabled.
2. Python users can do their other works (e.g., model init) between invoking init_process_group and their first collective call.
3. c10D would guarantee/wait for communicators to be initialized before issuing the first collective call.
4. For NCCL collective calls, the contract between python users and c10d is not changed much from blocking calls (C10d would wait the NCCL call to be ncclSuccess, or timeout, whichever happens first).


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang committed:
[c10d] allow nonblocking wrap of ncclCommInitRankConfig
Summary:
Different from other NCCL calls, we would like comms initialization to
return fast while initializaiton is in progress. So added a Macro to
simply check the call is inProgress and return immediately. This could
potentially let user thread to do its other works while NCCL is
initializing.

There is a ceveat: ProcessGroupNCCL seems to use lazy initialization and
initilize comms only when the first collective being called? In which
case we need to see if it helps user thread to speed up initialization.
Test Plan:
Turn on non-blocking mode flag, and run NCCL unit tests:
python test/distributed/test_c10d_nccl.py ProcessGroupNCCLTest
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wconstab commented:
glancing at your change, i think you're right to suspect this change won't optimize user code as much as we want, bc commInitRank won't be called early enough in the user program to allow it to overlap with things.

kwen2501 commented:
True. But a serious user would probably enable the early-init mode (setting a device id when calling `init_process_group`) so that there is a maximal time gap between `dist.init_process_group` and the first collective.

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -86,6 +86,18 @@
   } while (0)
 
+// Macro to throw on a non-successful NCCL return value for NONBLOCKING calls.
+#define C10D_NCCL_CHECK_NONBLOCKING(cmd, failureReason)                       \
```
Defining a new macro is okay. A neater way would be just to reuse the `C10D_NCCL_CHECK` macro: in it, rely on `#ifdef NCCL_HAS_NONBLOCKING` to add the `ncclInProgress` check.

shuqiangzhang replied:
After a second pass after the eagerMode change, feel it is better to use a separate macro to make the logic in Macro simple: e.g., no relative long verdict statement like (if (use_non_blocking && result!= success && result !=inprogress) || (!use_non_blocking && result!= success) for better readability

kwen2501 replied:
Hmm, I don't think we'd need to check `use_non_blocking` here. We just check for `result != ncclSuccess && result != ncclInProgress`. As in, after NCCL 2.14, we allow both values as immediate return. But, personal preference only, no need to take it seriously.

kwen2501 reviewed (approved):
LGTM. See also my comment and check if code can be consolidated. Thanks!

kwen2501 commented:
> True. But a serious user would probably enable the early-init mode (setting a device id when calling `init_process_group`) so that there is a maximal time gap between `dist.init_process_group` and the first collective.

Actually, we can use whether early-init is on as a guide of how we take advantage of the non-blocking mode.

If early-init is on:
    max cpu pass through (as this PR does)
else:
    max fault tolerence (as before this PR)

shuqiangzhang committed:
Update on "[c10d] allow nonblocking wrap of ncclCommInitRankConfig"


Summary:
Different from other NCCL calls, we would like comms initialization to
return fast while initializaiton is in progress. So added a Macro to
simply check the call is inProgress and return immediately. This could
potentially let user thread to do its other works while NCCL is
initializing.

There is a ceveat: ProcessGroupNCCL seems to use lazy initialization and
initilize comms only when the first collective being called? In which
case we need to see if it helps user thread to speed up initialization.
Test Plan:
Turn on non-blocking mode flag, and run NCCL unit tests:
python test/distributed/test_c10d_nccl.py ProcessGroupNCCLTest
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

kwen2501 commented:
> So In eager init mode, Is that user's responsibility check if the comm init is ncclSuccess?

Would be nice to hold CPU for them upon the first NCCL call.

kwen2501 commented:
Also having some discussion with NCCL team here:
https://github.com/NVIDIA/nccl/issues/1156#issuecomment-1910589514

shuqiangzhang commented:
I would hold this PR for now, until we have a UT to showcase how user can interact with the eager init

shuqiangzhang committed:
Update on "[c10d] allow nonblocking wrap of ncclCommInitRankConfig"


Summary:
Different from other NCCL calls, we would like comms initialization to
return fast while initializaiton is in progress. So added a Macro to
simply check the call is inProgress and return immediately. This could
potentially let user thread to do its other works while NCCL is
initializing.

There is a ceveat: ProcessGroupNCCL seems to use lazy initialization and
initilize comms only when the first collective being called? In which
case we need to see if it helps user thread to speed up initialization.
Test Plan:
Turn on non-blocking mode flag, and run NCCL unit tests:
python test/distributed/test_c10d_nccl.py ProcessGroupNCCLTest
Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
Updated the PR with the following intentions:
1. identify eagerMode init (as opposed to lazy init), in which case we will create NCCL comms without guarantees that they are fully initialized if NONBLOCKING mode is also enabled.
2. Python users can do their other works (e.g., model init) between invoking init_process_group and their first collective call.
3. c10D would guarantee/wait for communicators to be initialized before issuing the first collective call.
4. For NCCL collective calls, the contract between python users and c10d is not changed much from blocking calls (C10d would wait the NCCL call to be ncclSuccess, or timeout, whichever happens first).

Added an unit test, and printed out the order of events happened during initialization:

python test/distributed/test_c10d_nccl.py ProcessGroupNCCLTest.test_eager_mode_init

[W NCCLUtils.cpp:121] Warning: Using experimental non-blocking NCCL communicator. (function operator())
[W NCCLUtils.hpp:271] Rank 0: creating NCCL communicator eagerly in nonblocking mode
NCCL version 2.19.3+cuda12.0
[W NCCLUtils.cpp:121] Warning: Using experimental non-blocking NCCL communicator. (function operator())
[W NCCLUtils.hpp:271] Rank 1: creating NCCL communicator eagerly in nonblocking mode
[rank0]:[W ProcessGroupNCCL.cpp:2404] Running collective with sequence number 1 and with type: ALLREDUCE
[rank0]:[W NCCLUtils.cpp:52] Rank 0: NCCL communicator is initialized.
[rank1]:[W ProcessGroupNCCL.cpp:2404] Running collective with sequence number 1 and with type: ALLREDUCE
[rank1]:[W NCCLUtils.cpp:52] Rank 1: NCCL communicator is initialized.
.
----------------------------------------------------------------------
Ran 1 test in 17.209s
OK

shuqiangzhang commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -1357,6 +1357,22 @@ def test_comm_split_subgroup(self):
         self.assertEqual(backend.comm_split_count(), 1)
         self.assertEqual(tensor, original_tensor)
 
+    @requires_nccl_version((2, 18), "Need NCCL 2.18+ for ncclCommSplit")
```
will update the decorator here in the final diff ncclCommSplit is not needed here

chipturner replied:
maybe break this into two tests, one for eager, one for eager async?  the eager is implicitly tested in `test_comm_split_subgroup` but separating it out from the actual split testing would help if something actually broke the eager init directly

maybe better since we have a full test matrix of (eager_or_lazy, blocking_or_nonblocking) that we test all the combos in a single test case?

chipturner commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -26,9 +26,36 @@ ncclComm_t NCCLComm::getNcclComm() {
   return ncclComm_;
 }
 
+void NCCLComm::waitUtilInitialized(int timeoutSecs) {
```
this seems pretty similar to `ProcessGroupNCCL::WorkNCCL::synchronizeInternal` but I don't think that code would be easy to reuse or refactor.  But there are some things perhaps to borrow from it (like sleeping rather than busy looping).

chipturner commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -247,23 +261,40 @@ class NCCLComm {
-              &(comm->ncclComm_), numRanks, commId, rank, &config),
-          comm->ncclComm_,
-          c10::nullopt);
+      if (eagerMode) {
```
this is getting gnarly with the nested conditionals.  could it be restructured something like `if (eagerMode && nonblocking) { ... } else { ... }` to at least coalesce the two non-blocking inits into one?

shuqiangzhang replied:
after some thought,  we don't really need to differentiate if eagerMode is enabled for non-blocking mode. As long as nonblocking mode is enabled, we always init comms asynchronously. and it is each collective's duty to make sure its comm is initialized. So we don't need 4 logical branches

kwen2501 replied:
We can decouple the two conditions:
```
if (nccl_use_nonblocking()) {
  config.blocking = 0;
}

if (eagerMode) {
  C10D_NCCL_CHECK_NONBLOCKING(
    ncclCommInitRankConfig(...)
  )
} else {
  C10D_NCCL_CHECK_TIMEOUT(
    ncclCommInitRankConfig(...)
  )
  isInitialized = true;
}
```
That is, it should be safe to use `C10D_NCCL_CHECK_TIMEOUT` even if blocking mode is used (it will not enter the while loop in the macro).

chipturner commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1812,7 +1812,8 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
     int p2pRank,
-    bool isSendRecvSelf) {
+    bool isSendRecvSelf,
+    bool eagerMode) {
```
this feels like it should be earlier in the list of parameters, maybe before or after `opType`; can we change this API since it's internal?  it also would let us make it explicit rather than having a default value which seems nicer

shuqiangzhang replied:
After removing the eager mode branch in the new version, we wont need this parameter anymore

chipturner commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1969,7 +1971,13 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
   if (!nccl_use_nonblocking()) {
```
for readability I suggest swapping so the nonblocking case is first

chipturner commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -442,6 +479,8 @@ class NCCLComm {
 
  protected:
+  // a helper function to wait until the communicator is initialized;
+  void waitUtilInitialized(int timeoutSecs);
```
typo: `Util`

chipturner commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -366,6 +398,11 @@ class NCCLComm {
     return aborted_;
   }
 
+  bool isInitialized() const {
```
is this used in another PR or is it extraneous?

shuqiangzhang replied:
removed

chipturner reviewed (changes requested):
I can't speak much to the soundness of how nonblocking works in general (I lack the expertise in that area) but this seems reasonable overall.  Some added logging might be nice for when the init completes since it may generate delayed errors and be confusing otherwise.

shuqiangzhang committed:
Update on "[c10d] allow nonblocking wrap of ncclCommInitRankConfig"


resolve #117749 

Summary:
Updated the PR with the following intentions:

1. identify eagerMode init (as opposed to lazy init), in which case we will create NCCL comms without guarantees that they are fully initialized if NONBLOCKING mode is also enabled.
2. Python users can do their other works (e.g., model init) between invoking init_process_group and their first collective call.
3. c10D would guarantee/wait for communicators to be initialized before issuing the first collective call.
4. For NCCL collective calls, the contract between python users and c10d is not changed much from blocking calls (C10d would wait the NCCL call to be ncclSuccess, or timeout, whichever happens first).


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

kwen2501 reviewed (approved):
Taking a second look. LGTM. Minor comments re code simplification.

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -26,9 +31,38 @@ ncclComm_t NCCLComm::getNcclComm() {
+    auto timeElapsed = std::chrono::duration_cast<std::chrono::seconds>(
+                           currentTimepoint - startTimepoint)
+                           .count();
+    if (timeElapsed > timeoutSecs) {
+      std::string err = "NCCL timeout in communicator initialization.";
+      TORCH_CHECK_WITH(DistBackendError, false, err);
+    }
+    if (ncclComm_) {
+      ncclResult_t result;
+      ncclCommGetAsyncError(ncclComm_, &result);
+      if (result == ncclSuccess) {
+        LOG(INFO) << "Rank " << rank_ << ": NCCL communicator is initialized.";
+        initialized_ = true;
+      }
+    }
```
Strictly speaking, you'd need to put the `ncclCommGetAsyncError` check **BEFORE** the timeout check, to avoid a false positive timeout.

chipturner commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -28,31 +33,33 @@ ncclComm_t NCCLComm::getNcclComm() {
   }
   // only wait for initialization if nonblocking mode is enabled
   if (nccl_use_nonblocking()) {
```
I think it's worth checking `initialized_` here to avoid calling multiple functions on the happy (and most common) path.  This skips two `nccl_` calls as well as `steady_clock::now()`.  Overhead would be small but IMO a bit cleaner.

shuqiangzhang committed:
Update on "[c10d] allow nonblocking wrap of ncclCommInitRankConfig"


resolve #117749 

Summary:
Updated the PR with the following intentions:

1. identify eagerMode init (as opposed to lazy init), in which case we will create NCCL comms without guarantees that they are fully initialized if NONBLOCKING mode is also enabled.
2. Python users can do their other works (e.g., model init) between invoking init_process_group and their first collective call.
3. c10D would guarantee/wait for communicators to be initialized before issuing the first collective call.
4. For NCCL collective calls, the contract between python users and c10d is not changed much from blocking calls (C10d would wait the NCCL call to be ncclSuccess, or timeout, whichever happens first).


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1712,7 +1712,12 @@ std::exception_ptr ProcessGroupNCCL::checkForNCCLErrorsInternal(
     // When nonblocking mode is enabled by TORCH_NCCL_USE_COMM_NONBLOCKING,
     // ncclInProgress could be returned when there are pending NCCL calls.
     // In this case, no exception should be thrown
+    #ifdef NCCL_HAS_COMM_NONBLOCKING
```
This is a fix to the previous PR that broke fb internal tests (reverted).

shuqiangzhang committed:
Update on "[c10d] allow nonblocking wrap of ncclCommInitRankConfig"


resolve #117749 

Summary:
Updated the PR with the following intentions:

1. identify eagerMode init (as opposed to lazy init), in which case we will create NCCL comms without guarantees that they are fully initialized if NONBLOCKING mode is also enabled.
2. Python users can do their other works (e.g., model init) between invoking init_process_group and their first collective call.
3. c10D would guarantee/wait for communicators to be initialized before issuing the first collective call.
4. For NCCL collective calls, the contract between python users and c10d is not changed much from blocking calls (C10d would wait the NCCL call to be ncclSuccess, or timeout, whichever happens first).


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] allow nonblocking wrap of ncclCommInitRankConfig"


resolve #117749 

Summary:
Updated the PR with the following intentions:

1. identify eagerMode init (as opposed to lazy init), in which case we will create NCCL comms without guarantees that they are fully initialized if NONBLOCKING mode is also enabled.
2. Python users can do their other works (e.g., model init) between invoking init_process_group and their first collective call.
3. c10D would guarantee/wait for communicators to be initialized before issuing the first collective call.
4. For NCCL collective calls, the contract between python users and c10d is not changed much from blocking calls (C10d would wait the NCCL call to be ncclSuccess, or timeout, whichever happens first).


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge
